File: Java/src/main/java/com/DSA/Java/Graphs/SSSPP/SSSPPUsingList.java
package com.DSA.Java.Graphs.SSSPP;
import java.util.LinkedList;
import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyList.Graph;
import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyList.GraphNode;

public class SSSPPUsingList extends Graph {
    private void pathPrint(GraphNode node) {
        if (node.parent != null) {
            pathPrint(node.parent);
        } 
        System.out.print(node.name + " ");
    }

    private void bfsForSSSPPHelper(GraphNode root) {
        LinkedList<GraphNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            GraphNode currentNode = queue.remove(0);
            currentNode.isVisisted = true;
            System.out.print("Printing the path for " + currentNode.name + ": ");
            pathPrint(currentNode);
            System.out.println();

            for (GraphNode node : currentNode.neighbours) {
                if (!node.isVisisted) {
                    queue.add(node);
                    node.isVisisted = true;
                    node.parent = currentNode;
                }
            }
        }
    }

    public void bfsForSSSPP() {
        bfsForSSSPPHelper(getNodeList().get(0));
    }
}

File: C++/Graph/src/SSSPP/SSSPPUsingList.cpp
#include "../../include/SSSPP/SSSPPUsingList.h"
#include <queue>

void SSSPPUsingList::print_path(GraphNode *node) {
  if (node->parent != nullptr) {
    print_path(node->parent);
  }
  std::cout << node->name << " ";
}

void SSSPPUsingList::bfs_for_SSSPP_helper(GraphNode *root) {
  std::queue<GraphNode *> queue;
  if (root == nullptr) return ;
  queue.push(root);

  while (!queue.empty()) {
    GraphNode *current_node = queue.front();
    current_node->is_visited = true;
    queue.pop();

    std::cout << "Printing path for " << current_node->name << ": ";
    print_path(current_node);
    std::cout << "\n";

    std::vector<GraphNode *> neighbours = current_node->neighbours;
    for (const auto &neighbour : neighbours) {
      if (!neighbour->is_visited) {
        neighbour->is_visited = true;
        queue.push(neighbour);
        neighbour->parent = current_node;
      }
    }
  }
}

void SSSPPUsingList::bfs_for_SSSPP() {
  std::cout << "\n";
  bfs_for_SSSPP_helper(get_node_list()[0]);
}
